### DISCUSSION TITLE
How to properly check if a variable is PolyLike?

### DISCUSSION THREAD

**Discussion by iyanmv** [mhostetter/galois#496] [Q&A] [answered]
Hello,

How can we check if a variable is `PolyLike`?

First I tried to wrap my code with a try statement.

```python
import galois

try:
    poly = galois.Poly(provided_poly_like)

except ValueError, TypeError:
    pass
```

But this fails if `provided_poly_like` is already a `Poly`. For example,

```python
>>> galois.Poly(galois.conway_poly(2, 12))
TypeError: Argument 'coeffs' must be an instance of (<class 'list'>, <class 'tuple'>, <class 'numpy.ndarray'>, <class 'galois.Array'>), not <class 'galois.Poly'>.
```

So later I tried using `isinstance` and I found some stackoverflow questions about checking Union types at runtime, but probably I'm doing something wrong. What I tried is:

```python
assert isinstance(provided_poly_like, galois.typing.PolyLike.__args__)
```

**Comment by mhostetter** [comment, marked as answer]
I have a private constructor that converts `PolyLike` into `Poly`. I found it useful in the internal library code. At one point I considered making it a public constructor. I also toyed with the idea of making `galois.Poly()` accept `PolyLike`.

What are you thoughts? Is this what you're looking for? By the way, the type checking if very rudimentary. I've been looking for an easier way to runtime type check using type annotations, but every solution I've found has been shoddy.

https://github.com/mhostetter/galois/blob/a140a468fa1f7619f94ad2551f9c14e684ee3a34/src/galois/_polys/_poly.py#L130-L149

**Reply by iyanmv** [reply]
Yes, exactly what I was looking for, thanks! I missed it when searching the code :sweat_smile: 

And yes, I think it would be nice idea to make `galois.Poly()` accept `PolyLike`, or at least `Poly` in addition to what is already supported.

I would expect these two examples to work in the same way.

```python
from galois import GF2, Poly

array = GF2.Random(10)
new_array = GF2(array)  # Ok, we get a copy

poly = Poly.Str("x + 1")
new_poly = Poly(poly)  # This fails, but I think it should be Ok as well
```

**Reply by mhostetter** [reply]
Yeah, that's easy enough to do. If we make `Poly()` accept `PolyLike`, then we should probably remove `Poly.Int()` and `Poly.Str()`. If you'd like to write up an issue, then we can track it.
